### PR TITLE
alloc: make system runtime heap core specific

### PR DESCRIPTION
**Pull request by tlauda**
Extracts system runtime heap from common runtime heaps
and makes it core specific.

Signed-off-by: Tomasz Lauda <tomasz.lauda@linux.intel.com>

### CODE REVIEW

**Comment by lgirdwood** [comment]
@tlauda looks like a CI issue, Btw I have some heap updates coming too that help with multiple heaps and debug....

File: src/lib/alloc.c [hunk 1]
```diff
@@ -389,53 +406,47 @@ static void free_block(void *ptr)
...
-		map = cache_to_uncache(&heap->map[i]);
+	ptr = get_ptr_from_heap(cpu_heap, zone, caps, bytes);
 
-		/* is block big enough */
-		if (map->block_size < bytes)
-			continue;
+	/* other core should have the latest value */
+	if (core != cpu_get_id())
+		dcache_writeback_invalidate_region(cpu_heap,
+						   sizeof(*cpu_heap));
```
[the remark this answers is not in the document]
Comment (tlauda) [reply src/lib/alloc.c]: It's in the case there will be a need to make allocation for other core, like with the system heap and master core allocating space for slave cores before the boot.

File: src/platform/haswell/broadwell.x.in [hunk 2]
```diff
@@ -489,14 +489,6 @@ SECTIONS
     _system_heap_end = ABSOLUTE(.);
   } >system_heap :system_heap_phdr
```
[the remark this answers is not in the document]
Comment (tlauda) [reply src/platform/haswell/broadwell.x.in]: Fixed.

File: src/platform/cannonlake/include/platform/memory.h [hunk 3]
```diff
@@ -275,17 +280,40 @@
...
+#define HEAP_SYS_RUNTIME_1_BASE	(HEAP_SYS_RUNTIME_0_BASE + \
+				HEAP_SYS_RUNTIME_0_SIZE)
+#define HEAP_SYS_RUNTIME_1_SIZE \
+	(HEAP_SYS_RT_X_COUNT64 * 64 + HEAP_SYS_RT_X_COUNT512 * 512 + \
+	HEAP_SYS_RT_X_COUNT1024 * 1024)
+
+#define HEAP_SYS_RUNTIME_2_BASE	(HEAP_SYS_RUNTIME_1_BASE + \
+				HEAP_SYS_RUNTIME_1_SIZE)
+#define HEAP_SYS_RUNTIME_2_SIZE \
+	(HEAP_SYS_RT_X_COUNT64 * 64 + HEAP_SYS_RT_X_COUNT512 * 512 + \
```
[the remark this answers is not in the document]
Comment (tlauda) [reply src/platform/cannonlake/include/platform/memory.h]: It follows the same pattern as system heaps. Makes it more clear how many heaps there actually are. Let's keep it this way for now.

File: src/lib/alloc.c [hunk 4]
```diff
@@ -271,57 +271,74 @@ static struct mm_heap *get_heap_from_ptr(void *ptr)
...
 	uint32_t mask;
 	int i;
 
 	/* find first heap that support type */
-	for (i = 0; i < PLATFORM_HEAP_RUNTIME; i++) {
-		heap = cache_to_uncache(&memmap.runtime[i]);
-		mask = heap->caps & caps;
+	for (i = 0; i < count; i++) {
+		mask = heap[i].caps & caps;
 		if (mask == caps)
```
[the remark this answers is not in the document]
Comment (tlauda) [reply src/lib/alloc.c]: This function is only called by master core as slave cores are only using system and system runtime heaps.

**Comment by tlauda** [comment]
@lgirdwood I can't really get the failure. Host platform builds successfully with travis.
@libinyang RZONE_SYS_RUNTIME is for runtime objects allocated from kernel part of the code e.g. drivers, devices, interrupts. These resources are system specific, but also need to be freed. With the current implementation of the pipeline execution on slave cores, there should be no need for slave cores to allocate anything on runtime and buffer heaps. Everything goes through master core and is flushed to SRAM when needed.

File: src/lib/alloc.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (libinyang) [reply src/lib/alloc.c]: In this case, it should be not system_runtime, just system? From the source code, I see the core will always be the same cpu_get_id(), as it is only called in this file: ptr = rmalloc_sys(zone, caps, cpu_get_id(), bytes);

Anyway, it only impacts on the performance. I'm OK with check.

**Comment by libinyang** [comment]
> @lgirdwood I can't really get the failure. Host platform builds successfully with travis.
> @libinyang RZONE_SYS_RUNTIME is for runtime objects allocated from kernel part of the code e.g. drivers, devices, interrupts. These resources are system specific, but also need to be freed. With the current implementation of the pipeline execution on slave cores, there should be no need for slave cores to allocate anything on runtime and buffer heaps. Everything goes through master core and is flushed to SRAM when needed.

Thanks for clarification.

**Pull request merged**